File: src/main/java/net/hectus/neobb/shop/util/Items.java
package net.hectus.neobb.shop.util;

import net.kyori.adventure.text.Component;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.event.inventory.ClickType;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.inventory.ItemStack;
import org.jetbrains.annotations.NotNull;
import xyz.xenondevs.inventoryaccess.component.AdventureComponentWrapper;
import xyz.xenondevs.invui.gui.PagedGui;
import xyz.xenondevs.invui.gui.ScrollGui;
import xyz.xenondevs.invui.item.ItemProvider;
import xyz.xenondevs.invui.item.builder.ItemBuilder;
import xyz.xenondevs.invui.item.builder.SkullBuilder;
import xyz.xenondevs.invui.item.impl.AbstractItem;
import xyz.xenondevs.invui.util.MojangApiUtils;

import java.io.IOException;
import java.util.function.BiConsumer;

public class Items {
    public static final ItemStack WHITE_BACKGROUND = new net.hectus.neobb.shop.util.ItemBuilder(Material.WHITE_STAINED_GLASS).name(Component.empty()).build();
    public static final ItemStack GRAY_BACKGROUND = new net.hectus.neobb.shop.util.ItemBuilder(Material.GRAY_STAINED_GLASS_PANE).name(Component.empty()).build();
    public static final ItemStack BLACK_BACKGROUND = new net.hectus.neobb.shop.util.ItemBuilder(Material.BLACK_STAINED_GLASS_PANE).name(Component.empty()).build();

    public static class ClickItem extends AbstractItem {
        protected final ItemStack item;
        protected final BiConsumer<Player, InventoryClickEvent> clickConsumer;

        public ClickItem(ItemStack item, BiConsumer<Player, InventoryClickEvent> clickConsumer) {
            this.item = item;
            this.clickConsumer = clickConsumer;
        }

        @Override
        public ItemProvider getItemProvider() {
            return new ItemBuilder(item);
        }

        @Override
        public void handleClick(@NotNull ClickType clickType, @NotNull Player player, @NotNull InventoryClickEvent event) {
            clickConsumer.accept(player, event);
        }
    }
    
    public static class ScrollItem extends xyz.xenondevs.invui.item.impl.controlitem.ScrollItem {
        protected final boolean up;

        public ScrollItem(boolean up) {
            super(up ? -1 : 1);
            this.up = up;
        }

        @Override
        public ItemProvider getItemProvider(ScrollGui<?> gui) {
            try {
                return new SkullBuilder("MHF_Arrow" + (up ? "Up" : "Down")).setDisplayName(new AdventureComponentWrapper(Component.text("Scroll " + (up ? "Up" : "Down"))));
            } catch (MojangApiUtils.MojangApiException | IOException e) {
                throw new RuntimeException(e);
            }
        }
    }

    public static class PageItem extends xyz.xenondevs.invui.item.impl.controlitem.PageItem {
        protected final boolean forward;

        public PageItem(boolean forward) {
            super(forward);
            this.forward = forward;
        }

        @Override
        public ItemProvider getItemProvider(PagedGui<?> gui) {
            try {
                return new SkullBuilder("MHF_Arrow" + (forward ? "Right" : "Left")).setDisplayName(new AdventureComponentWrapper(Component.text((forward ? "Next" : "Last") + " Page")));
            } catch (MojangApiUtils.MojangApiException | IOException e) {
                throw new RuntimeException(e);
            }
        }
    }
}

File: src/main/kotlin/net/hectus/neobb/shop/util/Items.kt
package net.hectus.neobb.shop.util

import com.marcpg.libpg.util.ItemBuilder
import net.hectus.neobb.NeoBB
import net.kyori.adventure.text.Component
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.event.inventory.ClickType
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.inventory.ItemStack
import xyz.xenondevs.inventoryaccess.component.AdventureComponentWrapper
import xyz.xenondevs.invui.gui.PagedGui
import xyz.xenondevs.invui.gui.ScrollGui
import xyz.xenondevs.invui.item.ItemProvider
import xyz.xenondevs.invui.item.ItemWrapper
import xyz.xenondevs.invui.item.builder.SkullBuilder
import xyz.xenondevs.invui.item.impl.AbstractItem

object Items {
    val WHITE_BACKGROUND: ItemStack = ItemBuilder(Material.WHITE_STAINED_GLASS_PANE).name(Component.empty()).build()
    val GRAY_BACKGROUND: ItemStack = ItemBuilder(Material.GRAY_STAINED_GLASS_PANE).name(Component.empty()).build()
    val BLACK_BACKGROUND: ItemStack = ItemBuilder(Material.BLACK_STAINED_GLASS_PANE).name(Component.empty()).build()

    class ClickItem(private val item: ItemStack, private val clickConsumer: (Player, InventoryClickEvent) -> Unit) : AbstractItem() {
        override fun getItemProvider(): ItemProvider {
            return ItemWrapper(item)
        }

        override fun handleClick(clickType: ClickType, player: Player, event: InventoryClickEvent) {
            clickConsumer.invoke(player, event)
        }
    }

    class ScrollItem(private val up: Boolean) : xyz.xenondevs.invui.item.impl.controlitem.ScrollItem(if (up) -1 else 1) {
        override fun getItemProvider(gui: ScrollGui<*>): ItemProvider {
            try {
                return SkullBuilder("MHF_Arrow" + (if (up) "Up" else "Down")).setDisplayName(AdventureComponentWrapper(Component.text("Scroll " + (if (up) "Up" else "Down"))))
            } catch (e: Exception) {
                NeoBB.LOG.warn("Could not create skull for scroll item.", e)
                return ItemProvider.EMPTY
            }
        }
    }

    class PageItem(private val forward: Boolean) : xyz.xenondevs.invui.item.impl.controlitem.PageItem(forward) {
        override fun getItemProvider(gui: PagedGui<*>): ItemProvider {
            try {
                return SkullBuilder("MHF_Arrow" + (if (forward) "Right" else "Left")).setDisplayName(AdventureComponentWrapper(Component.text((if (forward) "Next" else "Last") + " Page")))
            } catch (e: Exception) {
                NeoBB.LOG.warn("Could not create skull for page item.", e)
                return ItemProvider.EMPTY
            }
        }
    }
}